linked-list.js

```diff
@@ -14,62 +14,188 @@ class LinkedList {
     this.head = null;
     this.tail = null;
     this.length = 0;
-
+    this.vals = vals
     for (let val of vals) this.push(val);
+
   }
 
   /** push(val): add new value to end of list. */
 
   push(val) {
-
+    let newNode = new Node(val)
+    if (this.head === null) {
+      this.head = newNode
+      this.tail = newNode
+    } else {
+      this.tail.next = newNode
+      this.tail = newNode
+    }
+    this.length++
   }
 
   /** unshift(val): add new value to start of list. */
 
   unshift(val) {
+    let newNode = new Node(val)
+    if (this.head === null) {
+      this.head = newNode
+      this.tail = newNode
+    } else {
+      newNode.next = this.head
+      this.head = newNode
+    }
+    this.length++
 
   }
 
   /** pop(): return & remove last item. */
 
   pop() {
-
+    let current, prevNode
+    if (this.head === null) { return null }
+    else if (this.head.next === null) {
+      current = this.head
+      this.head = null
+      this.tail = null
+      this.length--
+      return current.val
+    } else {
+      current = this.head
+      while (current.next !== this.tail) {
+        current = current.next
+      }
+      prevNode = this.tail
+      this.tail = current
+      current.next = null
+      this.length--
+      return prevNode.val
+    }
   }
 
   /** shift(): return & remove first item. */
 
   shift() {
+    let current, nextNode
+    if (this.head === null) { return null }
+    else if (this.head.next === null) {
+      current = this.head
+      this.head = null
+      this.tail = null
+      this.length--
+      return current.val
+    } else {
+      current = this.head
+      nextNode = this.head.next
+      this.head = nextNode
+      current.next = null
+      this.length--
+      return current.val
+    }
 
   }
 
   /** getAt(idx): get val at idx. */
 
   getAt(idx) {
-
+    if (idx >= this.length || idx < 0) {
+      throw new Error("Invalid index.");
+    }
+    return this.vals[idx]
```

Comment:
naahhh. this means your linkedlist is just another array. I'd iterate with a while loop and return the value there.